external/storm-kafka-new-consumer-api/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java

```diff
@@ -0,0 +1,457 @@
+    }
+
+    private void emitTuples(ConsumerRecords<K, V> consumerRecords) {
+        for (TopicPartition tp : consumerRecords.partitions()) {
+            final Iterable<ConsumerRecord<K, V>> records = consumerRecords.records(tp.topic());     // TODO Decide if want to give flexibility to emmit/poll either per topic or per partition
+            for (ConsumerRecord<K, V> record : records) {
+                final Values tuple = tupleBuilder.buildTuple(record);
+                final org.apache.storm.kafka.spout.MessageId messageId = new org.apache.storm.kafka.spout.MessageId(record);                                  // TODO don't create message for non acking mode. Should we support non acking mode?
+                collector.emit(kafkaSpoutStream.getStreamId(), tuple, messageId);           // emits one tuple per record
+                emittedTuples.put(messageId, tuple);
+                LOG.info("HMCL - Emitted tuple for record {}", record);
+            }
+        }
+    }
+
+    private boolean retry() {
+        return failed.size() > 0;
+    }
+
+    private void retryFailedTuples() {
+        for (TopicPartition tp : failed.keySet()) {
+            for (org.apache.storm.kafka.spout.MessageId msgId : failed.get(tp)) {
+                if (isInBlackList(msgId)) {
+                    removeFromBlacklist(msgId);
+                    removeFromFailed(tp, msgId);
+                } else {
+                    final Values tuple = emittedTuples.get(msgId);
+                    LOG.debug("Retrying tuple. [msgId={}, tuple={}]", msgId, tuple);
+                    collector.emit(kafkaSpoutStream.getStreamId(), tuple, msgId);
+                }
+            }
+        }
+    }
+
+    // all the tuples that are in traffic when the rebalance occurs will be added
+    // to black list to be disregarded when they are either acked or failed
+    private boolean isInBlackList(org.apache.storm.kafka.spout.MessageId msgId) {
+        return blackList.contains(msgId);
+    }
+
+    private void removeFromBlacklist(org.apache.storm.kafka.spout.MessageId msgId) {
+        blackList.remove(msgId);
+    }
+
+    // ======== Ack =======
+
+    @Override
+    public void ack(Object messageId) {
+        final org.apache.storm.kafka.spout.MessageId msgId = (org.apache.storm.kafka.spout.MessageId) messageId;
+        final TopicPartition tp = msgId.getTopicPartition();
+
+        if (isInBlackList(msgId)) {
+            removeFromBlacklist(msgId);
+        } else {
+            addAckedTuples(tp, msgId);
+            // Removed acked tuples from the emittedTuples data structure
+            emittedTuples.remove(msgId);
+            // if this acked msg is a retry, remove it from failed data structure
+            removeFromFailed(tp, msgId);
+        }
+    }
+
+    private void addAckedTuples(TopicPartition tp, org.apache.storm.kafka.spout.MessageId msgId) {
+        // lock because ack and commit happen in different threads
+        ackCommitLock.lock();
```

Comment:
`ackCommitLock` can be replaced with a synchronized block or method. There's no reason in this class to do explicit lock/unlock operations.